src/main/java/graphql/schema/SchemaUtil.java

```diff
@@ -224,52 +154,208 @@ private void collectTypesForInputObjects(GraphQLInputObjectType objectType, Map<
+
+        TRAVERSER.depthFirst(new TypeVisitorStub() {
+
+            @Override
+            public TraversalControl visitGraphQLObjectType(GraphQLObjectType node, TraverserContext<GraphQLType> context) {
+
+               node.replaceInterfaces(node.getInterfaces().stream()
+                        .map(type -> (GraphQLOutputType)typeMap.get(type.getName()))
+                        .collect(Collectors.toList()));
+                return super.visitGraphQLObjectType(node, context);
             }
-            if (type instanceof GraphQLUnionType) {
-                ((GraphQLUnionType) type).replaceTypeReferences(typeMap);
+
+            @Override
+            public TraversalControl visitGraphQLUnionType(GraphQLUnionType node, TraverserContext<GraphQLType> context) {
+
+                node.replaceTypes(node.getTypes().stream()
+                        .map(type -> (GraphQLOutputType)typeMap.get(type.getName()))
+                        .collect(Collectors.toList()));
+                return super.visitGraphQLUnionType(node, context);
             }
-        }
-    }
 
-    private void resolveTypeReferencesForFieldsContainer(GraphQLFieldsContainer fieldsContainer, Map<String, GraphQLType> typeMap) {
-        for (GraphQLFieldDefinition fieldDefinition : fieldsContainer.getFieldDefinitions()) {
-            fieldDefinition.replaceTypeReferences(typeMap);
-            for (GraphQLArgument argument : fieldDefinition.getArguments()) {
-                argument.replaceTypeReferences(typeMap);
+            @Override
+            public TraversalControl visitGraphQLTypeReference(GraphQLTypeReference node, TraverserContext<GraphQLType> context) {
+
+                final GraphQLType resolvedType = typeMap.get(node.getName());
+                Assert.assertTrue(resolvedType != null, "type %s not found in schema", node.getName());
+                context.getParentContext().thisNode().accept(context, new RefResolvingVisitor(resolvedType) );
+                return super.visitGraphQLTypeReference(node, context);
             }
-        }
-    }
+        },typeMap.values());
 
-    private void resolveTypeReferencesForInputFieldsContainer(GraphQLInputFieldsContainer fieldsContainer, Map<String, GraphQLType> typeMap) {
-        for (GraphQLInputObjectField fieldDefinition : fieldsContainer.getFieldDefinitions()) {
-            fieldDefinition.replaceTypeReferences(typeMap);
-        }
     }
 
+    @Deprecated
     GraphQLType resolveTypeReference(GraphQLType type, Map<String, GraphQLType> typeMap) {
         if (type instanceof GraphQLTypeReference || typeMap.containsKey(type.getName())) {
             GraphQLType resolvedType = typeMap.get(type.getName());
             Assert.assertTrue(resolvedType != null, "type %s not found in schema", type.getName());
             return resolvedType;
         }
-        if (type instanceof GraphQLList) {
-            ((GraphQLList) type).replaceTypeReferences(typeMap);
+        return type;
+    }
+
+
```

Comment:
There is enough visitor code here that it deserves its own file.

Perhaps not one class per visitor so I would create a package protected holder of these classes called say "TypeVisitors" that contained these classes